gapic-common/lib/gapic/rest/client_stub.rb

```diff
@@ -125,28 +125,84 @@ def make_put_request uri:, body: nil, params: {}, options: {}
+        # Converts hash and nil to an options object
+        options = ::Gapic::CallOptions.new(**options.to_h) unless options.is_a? ::Gapic::CallOptions
+        deadline = calculate_deadline options
+        retried_exception = nil
+        next_timeout = get_timeout deadline
+
+        begin
+          base_make_http_request(verb,
+                                 uri: uri,
+                                 body: body,
+                                 params: params,
+                                 metadata: options.metadata,
+                                 timeout: next_timeout,
+                                 is_server_streaming: is_server_streaming,
+                                 &block)
+        rescue ::Faraday::Error => e
+          next_timeout = get_timeout deadline
+
+          if next_timeout&.positive? && options.retry_policy.call(e)
+            retried_exception = e
+            retry
+          end
+
+          unless next_timeout&.positive?
+            raise Gapic::GRPC::DeadlineExceededError.new e.message, root_cause: retried_exception
+          end
+
+          raise e
+        end
+      end
+
+      private
+
+      ##
+      # @private
+      # Sends a http request via Faraday
+      # @param verb [Symbol] http verb
+      # @param uri [String] uri to send this request to
+      # @param body [String, nil] a body to send with the request, nil for requests without a body
+      # @param params [Hash] query string parameters for the request
+      # @param metadata [Hash] additional headers for the request
       # @param is_server_streaming [Boolean] flag if method is streaming
       # @yieldparam chunk [String] The chunk of data received during server streaming.
       # @return [Faraday::Response]
-      def make_http_request verb, uri:, body:, params:, options:, is_server_streaming: false
+      def base_make_http_request verb, uri:, body:, params:, metadata:, timeout:, is_server_streaming: false
         if @numeric_enums && (!params.key?("$alt") || params["$alt"] == "json")
           params = params.merge({ "$alt" => "json;enum-encoding=int" })
         end
-        options = ::Gapic::CallOptions.new(**options.to_h) unless options.is_a? ::Gapic::CallOptions
+
         @connection.send verb, uri do |req|
           req.params = params if params.any?
           req.body = body unless body.nil?
-          req.headers = req.headers.merge options.metadata
-          req.options.timeout = options.timeout if options.timeout&.positive?
+          req.headers = req.headers.merge metadata
+          req.options.timeout = timeout if timeout&.positive?
           if is_server_streaming
             req.options.on_data = proc do |chunk, _overall_received_bytes|
               yield chunk
             end
           end
         end
       end
+
+      def calculate_deadline options
+        return if options.timeout.nil?
+        return if options.timeout.negative?
+
+        Time.now + options.timeout
```

Comment:
Since we're not actually using the deadline as a real clock time but just a way to calculate timeouts, better to use `Process.clock_gettime(Process::CLOCK_MONOTONIC)` instead of `Time.now`, and same in `get_timeout` below. This prevents the bug where a regular NTS update changes the system clock in the meantime, resulting in an incorrect timeout.

Reply:
:+1: